sdk/src/Core/Amazon.Runtime/Credentials/_bcl+coreclr/DefaultInstanceProfileAWSCredentials.cs

```diff
@@ -84,25 +84,14 @@ public override ImmutableCredentials GetCredentials()
         {
             CheckIsIMDSEnabled();
 
-            ImmutableCredentials credentials;
-            try
-            {
-                refreshLock.EnterReadLock();
-                credentials = lastRetrievedCredentials?.Copy();
-            }
-            finally
-            {
-                refreshLock.ExitReadLock();
-            }
+            var credentials = lastRetrievedCredentials?.Copy();
```

Comment:
IMHO, if lastRetrievedCredentials is null, it should call RenewCredentials(null) before returning.
Related  https://github.com/aws/aws-sdk-net/issues/1004#issuecomment-407611176

Reply:
@arishlabroo I have taken the other option suggested, don't always set the `lastRetrievedCredentials` on an exception. I am now doing

````
            catch (TaskCanceledException e)
            {
                // in this case, keep the lastRetrievedCredentials
            }
            catch (Exception e)
            {
                lastRetrievedCredentials = null;
               ````
````

There is also the option to _never_ set `lastRetrievedCredentials` to null on any exception. Or to try to `RenewCredentials` when it's null in `GetCredentials` as you suggest above.

I'm not quite sure which of these is the best way to avoid nulls.

Reply:
Thanks for this PR Anthony.  We have a CR review going within our team that contains some of the changes you've already addressed here.  So, if you are open to it let's merge the changes :)

We were thinking of addressing the thread contention problem by doing lazy fetch of credentials instead of doing the initial fetch at creation of the singleton instance. 

1. We will no longer do `_instance.RenewCredentials(null);`  inside the lock.
2. Start the timer immediately in the constructor `credentialsRetrieverTimer = new Timer(RenewCredentials, null, TimeSpan.Zero, neverTimespan);` 
3. Refactor the credential fetch logic in `RenewCredentials()`
4. In `GetCredentials()` we would do a fallback of directly fetching IMDS credentials using the refactored logic from (3.) if `lastRetrievedCredentials` is `null.  This will also address some cases of https://github.com/aws/aws-sdk-net/issues/1004 

Using the repro application your team provided, we saw only about 2~3 direct fetches from IMDS before clients started using cached credentials.  If the above description is not detailed enough, I can create PR--sometime later today--with the change so we can take a look at our proposed change.  

What do you think about this approach?

Reply:
This sounds great. You are of course welcome to merge this PR or just use parts as suggestions for other PRs.

So the issues that I have seen (and I think #1004) are not just at  thread contention at startup, they happen after the app has been running for a while, and load is elevated. `RenewCredentials` is where I think it goes wrong.

We have not yet had the time to load-test code using this PR, but hope to do that soon.

Reply:
I created a PR of our change here as a reference: https://github.com/aws/aws-sdk-net/pull/1017

If you could do a review & roll up the change in to this PR, we'll start the process of merging this PR in.  Before we do the merge, however, we would like to make sure that the change in https://github.com/aws/aws-sdk-net/pull/1017 does indeed alleviate the thread contention problem in your app. environment.

Reply:
I'm not quite clear - do you want me to review #1017 as it stands, or to merge #1017 into #1010 (this PR?)

Reply:
Sorry for the confusion!  I would like you to merge #1017 into this PR.